pkg/crcontrollers/csibmnode/controller.go

```diff
@@ -0,0 +1,328 @@
+		}).
+		Watches(&source.Kind{Type: &coreV1.Node{}}, &handler.EnqueueRequestForObject{}). // secondary resource
+		WithEventFilter(predicate.Funcs{
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				if _, ok := e.ObjectOld.(*nodecrd.CSIBMNode); ok {
+					return true
+				}
+
+				nodeOld, ok := e.ObjectOld.(*coreV1.Node)
+				if !ok {
+					return false
+				}
+				nodeNew := e.ObjectNew.(*coreV1.Node)
+
+				annotationAreTheSame := reflect.DeepEqual(nodeOld.GetAnnotations(), nodeNew.GetAnnotations())
+				addressesAreTheSame := reflect.DeepEqual(nodeOld.Status.Addresses, nodeNew.Status.Addresses)
+				bmc.log.Debugf("UpdateEvent for k8s node %s. Annotations are the same - %v. Addresses are the same - %v.",
+					nodeOld.Name, annotationAreTheSame, addressesAreTheSame)
+
+				return !annotationAreTheSame || !addressesAreTheSame
+			},
+		}).
+		Complete(bmc)
+}
+
+// Reconcile reconcile CSIBMNode CR and k8s CSIBMNode objects
+// at first define for which object current Reconcile is triggered and then run corresponding reconciliation method
+func (bmc *Controller) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ll := bmc.log.WithFields(logrus.Fields{
+		"method": "Reconcile",
+		"name":   req.Name,
+	})
+
+	var err error
+	// if name in request doesn't start with namePrefix controller tries to read k8s node object at first
+	// however if it get NotFound error it tries to read CSIBMNode object as well
+	if !strings.HasPrefix(req.Name, namePrefix) {
+		k8sNode := new(coreV1.Node)
+		err = bmc.k8sClient.ReadCR(context.Background(), req.Name, k8sNode)
+		switch {
+		case err == nil:
+			ll.Infof("Reconcile k8s node %s", k8sNode.Name)
+			return bmc.reconcileForK8sNode(k8sNode)
+		case !k8sError.IsNotFound(err):
+			ll.Errorf("Unable to read node object: %v", err)
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	// try to read CSIBMNode
+	bmNode := new(nodecrd.CSIBMNode)
+	err = bmc.k8sClient.ReadCR(context.Background(), req.Name, bmNode)
+	switch {
+	case err == nil:
+		ll.Infof("Reconcile CSIBMNode %s", bmNode.Name)
+		return bmc.reconcileForCSIBMNode(bmNode)
+	case !k8sError.IsNotFound(err):
+		ll.Errorf("Unable to read CSIBMNode object: %v", err)
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	ll.Warnf("unable to detect for which object (%s) that reconcile is. The object may have been deleted", req.String())
+	return ctrl.Result{}, nil
+}
+
+func (bmc *Controller) reconcileForK8sNode(k8sNode *coreV1.Node) (ctrl.Result, error) {
+	ll := bmc.log.WithFields(logrus.Fields{
+		"method": "reconcileForK8sNode",
+		"name":   k8sNode.Name,
+	})
+
+	if len(k8sNode.Status.Addresses) == 0 {
+		err := errors.New("addresses are missing for current k8s node instance")
+		ll.Error(err)
+		return ctrl.Result{Requeue: false}, err
+	}
+
+	// get corresponding CSIBMNode CR name from cache
+	var bmNode = &nodecrd.CSIBMNode{}
+	if bmNodeName, ok := bmc.cache.getCSIBMNodeName(k8sNode.Name); ok {
```

Comment:
Situation: 
1. k8s node1 created -> k8s node1 reconcile -> bmnode1 created
2. 5 minutes later 🧽
3. k8s node1 removed
4. another k8s node added with same name as k8s node1
If I understand correctly, we will read data from cache and update annotation without validation of ip address. Is it expected behavior?

Reply:
Cache just holds mapping between names. It's main purpose is to reduce amount of List operation. When we get name from cache then we read corresponding CR/node from k8s APIs.
In case that you described, controller reads corresponding node and CSIBMNode CR objects and compare their addresses. If addresses aren't the same, controller log error and do nothing.

Reply:
> In case that you described, controller reads corresponding node and CSIBMNode CR objects and compare their addresses. If addresses aren't the same, controller log error and do nothing.
I can't find where we check that IP addresses are same if read CSIBMNode CR name from cache.

Reply:
addresses in both places